webapp/certification/views.py

```diff
@@ -15,6 +15,99 @@
 )
 
 
+def certification_hardware_details(canonical_id, release):
+    models = api.certified_models(canonical_id=canonical_id)["objects"]
+
+    if not models:
+        abort(404)
+
+    model_devices = api.certified_model_devices(
+        canonical_id=canonical_id, limit="0"
+    )["objects"]
+    model_releases = api.certified_model_details(
+        canonical_id=canonical_id, limit="0"
+    )["objects"]
+
+    # Release section
+    release_details = next(
+        (
+            detail
+            for detail in model_releases
+            if detail["certified_release"] == release
+        ),
+        None,
+    )
+
+    if not release_details:
+        abort(404)
+
+    hardware_details = {}
+    for device in model_devices:
+        device_info = {
+            "name": (
+                f"{device['make']} {device['name']}"
+                f" {device['subproduct_name']}"
+            ),
+            "bus": device["bus"],
+            "identifier": device["identifier"],
+        }
+
+        category = device["category"]
+        if category not in ["BIOS", "USB"]:
+            category = category.capitalize()
+
+        if category not in hardware_details:
+            hardware_details[category] = []
+
+        hardware_details[category].append(device_info)
+
+    # Components section
+    components = {}
+    for device_category, devices in release_details.items():
+        if (
+            device_category in ["video", "processor", "network", "wireless"]
+            and devices
+        ):
+            device_category = device_category.capitalize()
+
+            components[device_category] = []
+
+            if device_category in components:
+                for device in devices:
+                    components[device_category].append(
+                        {
+                            "name": (
+                                f"{device['make']} {device['name']}"
+                                f" {device['subproduct_name']}"
+                            ),
+                            "bus": device["bus"],
+                            "identifier": device["identifier"],
+                        }
+                    )
+
+    # Build model name
+    model_names = next(model["model"] for model in models)
+
+    category = models[0]["category"]
+    # default to category, which contains the least specific form_factor
+    form_factor = release_details and release_details.get(
```

Comment:
No need to check for `release_details` was we abort every requests where it is falsy.

Reply:
Moved down. Done